flink-connectors/flink-connector-gcp-pubsub/src/main/java/org/apache/flink/streaming/connectors/gcp/pubsub/PubSubSource.java

```diff
@@ -0,0 +1,292 @@
+import com.google.pubsub.v1.PubsubMessage;
+import com.google.pubsub.v1.ReceivedMessage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.time.Duration;
+import java.util.List;
+import java.util.concurrent.CancellationException;
+
+import static com.google.cloud.pubsub.v1.SubscriptionAdminSettings.defaultCredentialsProviderBuilder;
+
+/**
+ * PubSub Source, this Source will consume PubSub messages from a subscription and Acknowledge them on the next checkpoint.
+ * This ensures every message will get acknowledged at least once.
+ */
+public class PubSubSource<OUT> extends RichSourceFunction<OUT>
+	implements ResultTypeQueryable<OUT>, ParallelSourceFunction<OUT>, CheckpointListener, ListCheckpointed<AcknowledgeIdsForCheckpoint<String>> {
+	public static final int NO_MAX_MESSAGES_TO_ACKNOWLEDGE_LIMIT = -1;
+	private static final Logger LOG = LoggerFactory.getLogger(PubSubSource.class);
+	protected final PubSubDeserializationSchema<OUT> deserializationSchema;
+	protected final PubSubSubscriberFactory pubSubSubscriberFactory;
+	protected final Credentials credentials;
+	protected final int maxMessagesToAcknowledge;
+	protected final AcknowledgeOnCheckpointFactory acknowledgeOnCheckpointFactory;
+
+	protected transient AcknowledgeOnCheckpoint<String> acknowledgeOnCheckpoint;
+	protected transient PubSubSubscriber subscriber;
+
+	protected transient volatile boolean isRunning;
+
+	PubSubSource(PubSubDeserializationSchema<OUT> deserializationSchema,
+				PubSubSubscriberFactory pubSubSubscriberFactory,
+				Credentials credentials,
+				int maxMessagesToAcknowledge,
+				AcknowledgeOnCheckpointFactory acknowledgeOnCheckpointFactory) {
+		this.deserializationSchema = deserializationSchema;
+		this.pubSubSubscriberFactory = pubSubSubscriberFactory;
+		this.credentials = credentials;
+		this.maxMessagesToAcknowledge = maxMessagesToAcknowledge;
+		this.acknowledgeOnCheckpointFactory = acknowledgeOnCheckpointFactory;
+	}
+
+	@Override
+	public void open(Configuration configuration) throws Exception {
+		super.open(configuration);
+		if (hasNoCheckpointingEnabled(getRuntimeContext())) {
+			throw new IllegalArgumentException("The PubSubSource REQUIRES Checkpointing to be enabled and " +
+				"the checkpointing frequency must be MUCH lower than the PubSub timeout for it to retry a message.");
+		}
+
+		getRuntimeContext().getMetricGroup().gauge("PubSubMessagesProcessedNotAcked", this::getOutstandingMessagesToAck);
+
+		createAndSetPubSubSubscriber();
+		this.isRunning = true;
+	}
+
+	private boolean hasNoCheckpointingEnabled(RuntimeContext runtimeContext) {
+		return !(runtimeContext instanceof StreamingRuntimeContext && ((StreamingRuntimeContext) runtimeContext).isCheckpointingEnabled());
+	}
+
+	@Override
+	public void run(SourceContext<OUT> sourceContext) throws Exception {
+		while (isRunning) {
+			try {
+				if (maxMessagesToAcknowledgeLimitReached()) {
```

Comment:
The situation we encountered was that we had a large at  dataskew at some point. This dataskew caused a couple of the subtasks to have loads of backpressure which in turn resulted in expired checkpoints.

This made it that the non backpressured subtasks kept pulling and processing more messages but no checkpoints would happening for a (long) time eventualy resulting in OOM exceptions.

Now we never put a profiler into the job to verify this a 100% so it could've been something else but the above could happen right?

Reply:
@Xeli Thanks for the explanation. In the case you mentioned, should the checkpoint timeout be set to longer, or should the CP interval be set to smaller?

To explain my concern, the motivation and behavior of the current code are following:
1. Continuous checkpoint failure may cause the number of unacknowledged messages to accumulate and grow.
2. To avoid OOM in situation 1, we need to cap the maximum unacknowledged messages.
3. With the while loop, the source may stall forever if checkpoints keep failing. After removing the while loop, the maximum unacknowledged messages will grow slower, i.e. consumption rate is throttled at ~10 pull request per sec, but continuous CP failure may still eventually cause OOM.

From what I see in the current Flink code, the design is following:
1. Ideally, checkpoints are performed asynchronously and should not slow down the processing.
2. When there are continuous checkpoint failures, it is a problem by itself, therefore the job can fail by user specifying max tolerable continuous CP failure in checkpoint config.
3. The flow control is done either by explicit throttling or implicitly via backpressure.

So while I understand the motivation here, the actual behavior seems not quite align with the overall design here.

Reply:
Thank you for the clear explaination.

Without this limit, we will run into OOM exceptions. Lowering the CP interval does not matter because it is always trying to checkpoint but they just take too long to complete. Increasing the checkpoint timeout would work, but if this is increased too far the source can still fill up the memory.

I like the idea of throttling via backpressure but with a dataskew, I don't know how the 'non-backpressured' subtasks can be throttled and explicit throttling is basically what this limit does right?


Ps. Regarding concern 3, the maximum acknowledged messages won't grow at ~10 pull req/s but stop at the limit because of the `continue`, right?

Reply:
Sorry I forgot to mention that in the earlier comments, by explicit rate limiting I actually meant using `FlinkConnectorRateLimiter` and expose corresponding configurations to the users. In general, we would like to avoid the case that each individual connector implement their own way to do the throttling.

> Without this limit, we will run into OOM exceptions. Lowering the CP interval does not matter because it is always trying to checkpoint but they just take too long to complete. Increasing the checkpoint timeout would work, but if this is increased too far the source can still fill up the memory.

I am curious what is the cause of such continuous slow checkpoint. Usually a long checkpoint is caused by large state to be saved. And because the state to be saved are likely has a positive correlation with the records processed, a smaller CP interval would reduce the state to be saved in each CP and thus make the it faster.

Reply:
1. I really like the explicit ratelimiting by using `FlinkConnectorRateLimiter` I didn't know this existed. However I don't feel comfortable adding this before the feature freeze (due to it pulling in guava, impact on performance, oom issues?). 

I propose we keep the limit as for now and add this FlinkConnectorRateLimiter to FLINK-13083

Reply:
It is perfectly fine to introduce the rate limiter later. The reason I am sticking to this is because `maxMessagesToAcknowledge` is a public API in the constructor. As long as we replace it with the rate limiter before Flink 1.9 officially releases, I am not concerned.

Reply:
I've added it to the new jira ticket 👍